### ISSUE TITLE
Add SSL documentation

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#11141]
|[<img alt="puzzlement's avatar" src="https://avatars.githubusercontent.com/u/94942?s=50" width="50" height="50">](https://github.com/puzzlement)| @puzzlement reported|
|-|-|
|Trac ID|trac#1627|
|Type|enhancement|
|Created|2006-04-09 01:44:52Z|
|Branch|https://github.com/twisted/twisted/tree/ssl-docs-1627|

We need a chapter in the user guide about how to use SSL in Twisted. Topics that should be covered:

 - listenSSL / connectSSL
 - startTLS
 - how to do server or client certificate verification

The original description of this ticket follows:

 - http://blog.vrplumber.com/356
 - http://blog.vrplumber.com/358
 - http://blog.vrplumber.com/371
 - http://blog.vrplumber.com/349
 - http://blog.vrplumber.com/310
 - http://www.vrplumber.com/programming/twistedlowlevel05.pdf

MIT licenced.

Attachments:

* [index.xhtml](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/3c7/3c7bf807f29cfe10e5cf0897570ded86d1e0a174/edbb94cdca4e7b1d1e2dd709ccca823d20174e31.xhtml) (5359 bytes) - added by jesstess on 2009-08-17 04:41:37Z - 
* [ssl.xhtml](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/3c7/3c7bf807f29cfe10e5cf0897570ded86d1e0a174/6a5e425f3f5406dc598bf9e02e75c480d9a302cc.xhtml) (11744 bytes) - added by jesstess on 2009-08-17 04:42:12Z - 
* [firstround_ssl.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/3c7/3c7bf807f29cfe10e5cf0897570ded86d1e0a174/610d5a96be88fd9b9c56615c993ab8ce6056fef7.patch) (10825 bytes) - added by jesstess on 2009-08-26 19:27:59Z - updates based on exarkun's first set of comments
* [secondround_ssl.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/3c7/3c7bf807f29cfe10e5cf0897570ded86d1e0a174/c11e30145cd7c2bc36873e04990c4951fdb9be52.patch) (13670 bytes) - added by jesstess on 2009-08-30 03:16:09Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__1627 1627
type__enhancement enhancement
reporter__hypatia hypatia
priority__normal normal
milestone__ 
branch__branches_ssl_docs_1627 branches/ssl-docs-1627
branch_author__jesstess jesstess
status__closed closed
resolution__fixed fixed
component__core core
keywords__documentation documentation
time__1144547092000000 1144547092000000
changetime__1251738479000000 1251738479000000
version__ 
owner__ 
cc__thijs cc__jesstess cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="puzzlement's avatar" src="https://avatars.githubusercontent.com/u/94942?s=50" width="50" height="50">](https://github.com/puzzlement)|@puzzlement set owner to edsuom|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_2"></a>|@thijstriemstra set owner to @thijstriemstra<br>@thijstriemstra set status to `assigned`|
|-|-|

Only [one of the links](http://blog.vrplumber.com/310) of the blog posts actually showed a code sample for me, and I suggest putting the link to the [PDF](http://www.vrplumber.com/programming/twistedlowlevel05.pdf) on the wiki, instead of converting it into a xhtml doc.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_3"></a>|@thijstriemstra removed owner<br>@thijstriemstra set status to `new`|
|-|-|

what do you think?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set owner to @thijstriemstra|
|-|-|

It seems like a lot of his content has been lost somehow.  I think the pages being served at those URLs are not what they used to be.

I think several of the posts represent important holes in our documentation.  For example, we really do have essentially zero SSL documentation.  Fixing that is extremely worthwhile.

I'm unexcited about linking to the PDF.  It has a lot of non-Twisted stuff in it and for all we know it will disappear like a lot of the other content has already.  It also mostly regurgitates stuff we do actually have documentation for.

So I guess what I'd like to see come out of this ticket is some SSL documentation.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_5"></a>|@thijstriemstra commented|
|-|-|

In that case I'm renaming the ticket.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_6"></a>|@thijstriemstra set status to `assigned`|
|-|-|

The [INSTALL](INSTALL) file could also use some notes on the SSL requirements..

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_8"></a>|@jesstess set owner to @jesstess<br>@jesstess set status to `new`|
|-|-|

thijs - I have toy examples for what's requested in the ticket and have time in the next few days to work on a writeup, so I'm going to reassign it to myself.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_9"></a>|@jesstess removed owner|
|-|-|

Attached are an SSL howto and a diff of the core howto index including it under the Tutorials bullet.

I also have documentation about generating certs for testing if people think that would be useful to include. 

I opted to put the full examples in the howto instead of linking to listings in examples/ since the basic SSL echo server and client are already represented there, but I can move them into examples/ if people would prefer that.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

(In [27234]) Branching to 'ssl-docs-1627'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

(In [27235]) Add ssl.xhtml and turn the attached index.xhtml into a patch and apply it

refs [#11141](https://github.com/twisted/twisted/issues/11141)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun commented|
|-|-|

This is a really awesome contribution.  Thank you very much for working on this ticket, jesstess. :)  We're long, long overdue on having these docs.

To facilitate further edits, I've added your changes to a branch.  Any further diffs should be generated relative to that branch.

Normally I would insert my review feedback here, with each point consisting of some directions on how to find the part of the docs I was commenting on and then an actual comment.  I've found this approach increasingly problematic (specifically for doc changes), so I'm going to try something else.  I'm going to edit the branch and insert markers where I think there's work to be done.  Hopefully this will be easier on both sides of the process.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun commented|
|-|-|

(In [27239]) Some initial comments

refs [#11141](https://github.com/twisted/twisted/issues/11141)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun set owner to @jesstess<br>@exarkun set status to `new`|
|-|-|

Changes added in r27239 (follow the link to see the diff, if you want).  Let me know if you have questions about anything.  Thanks again.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_16"></a>|@jesstess removed owner|
|-|-|

The new patch hopefully reflects exarkun's feedback. I know I need to fix line length and trailing white space issues but wanted to wait on that so it wouldn't obscure content changes.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun commented|
|-|-|

(In [27246]) Apply firstround_ssl.patch

refs [#11141](https://github.com/twisted/twisted/issues/11141)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_18"></a>|@exarkun set owner to @jesstess|
|-|-|

Thanks for omitting the cosmetic fixes from this patch.  It definitely helps with readability.  Lots of people don't consider that. :)

The changes to the text and markup in this new patch look great.  I've applied them to the branch (as visible above).  I don't have any more significant comments on the content of the document.  I think it's good to merge.  If you want to make some more cosmetic edits (line length, etc), go for it, and then I'll merge it.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_19"></a>|@jesstess commented|
|-|-|

Adjusted line width to 80 characters, removed trailing white space, and removed the isClient line from the startTLS server example since there's already been some confusion about it expressed on IRC.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)|@jesstess removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_21"></a>|@thijstriemstra set owner to @jesstess|
|-|-|

1.There's a line that goes like "is a convenience server class that subclasses ContextFactory and has default parameters to the SSL handshake and connection". Seems to me that the word 'has' should be replaced by 'adds'?
 1. There's a couple of references to ```ContextFactory``` that probably need ```<code></code>``` markup.
 1. Same for ```DefaultOpenSSLContextFactory```, ```ClientContextFactory```, ```ServerTLSContext``` and ```ClientTLSContext```
 1. In ```<a href="http://www.wireshark.org/">wireshark</a>``` the linked 'wireshark' should be 'Wireshark'.
 1. In the line 'its client, http, and xmlrpc modules.' those modules should either have ```<code></code>``` or ```<code class="API"``` markup

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_22"></a>|@exarkun commented|
|-|-|

(In [27261]) Apply secondround_ssl.patch

refs [#11141](https://github.com/twisted/twisted/issues/11141)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_23"></a>|@exarkun commented|
|-|-|

(In [27262]) mark up some more code, fix miswording and casing of wireshark

refs [#11141](https://github.com/twisted/twisted/issues/11141)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_24"></a>|@exarkun removed owner|
|-|-|

Thanks for the extra set of eyes, thijs. :)  I think I addressed all those points in r27262.  The text may need to be reflowed again, but I left off doing that to make the diff easier to read.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_25"></a>|@thijstriemstra set owner to @exarkun|
|-|-|

Cool, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_26"></a>|@exarkun commented|
|-|-|

(In [27264]) reflow a couple particularly long lines

refs [#11141](https://github.com/twisted/twisted/issues/11141)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_27"></a>|@exarkun set status to `closed`|
|-|-|

(In [27267]) Merge ssl-docs-1627

Author: jesstess
Reviewer: exarkun, thijs
Fixes: [#11141](https://github.com/twisted/twisted/issues/11141)

Add a howto-style document explaining Twisted's SSL features.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
